pkg/modeling/modeling.go

```diff
@@ -0,0 +1,341 @@
+	for _, rm := range resourceModels {
+		tmp := map[clusterapis.ResourceName]resource.Quantity{}
+		for _, rmItem := range rm.Ranges {
+			if len(rsName) != len(rm.Ranges) {
+				rsName = append(rsName, rmItem.Name)
+			}
+			tmp[rmItem.Name] = rmItem.Min
+		}
+		rsList = append(rsList, tmp)
+	}
+
+	if len(rsName) != 0 && len(rsList) != 0 && (len(rsName) != len(rsList[0])) {
+		return nil, errors.New("the number of resourceName is not equal the number of resourceList")
+	}
+	var rs ResourceSummary
+	if len(rsName) != 0 {
+		defaultModelSorting = rsName
+	}
+	rs = make(ResourceSummary, len(rsList))
+	// generate a sorted array by first priority of ResourceName
+	modelSortings = make([][]int64, len(rsList))
+	for index := 0; index < len(rsList); index++ {
+		for i, name := range rsName {
+			tmpQuantity := rsList[index][name]
+			quantityNum, ok := tmpQuantity.AsInt64()
+			if !ok {
+				klog.Infof("Unable to parse the values of %v's quantity in the cluster", rsName[0])
+			}
+			modelSortings[i] = append(modelSortings[i], quantityNum)
+		}
+	}
+	return rs, nil
+}
+
+// NewClusterResourceNode create new cluster resource node
+func NewClusterResourceNode(resourceList corev1.ResourceList) ClusterResourceNode {
+	return ClusterResourceNode{
+		quantity:     1,
+		resourceList: ConvertToResourceList(resourceList),
+	}
+}
+
+func (rs *ResourceSummary) getIndex(crn ClusterResourceNode) int {
+	index := math.MaxInt
+	for i, m := range defaultModelSorting {
+		tmpQuantity := crn.resourceList[m]
+		quantityNum, ok := tmpQuantity.AsInt64()
+		if !ok {
+			klog.Infof("Unable to parse the values of %v's quantity in the cluster", crn.resourceList)
+		}
+		tmpIndex := searchLastLessElement(modelSortings[i], quantityNum)
+
+		if tmpIndex < index {
+			index = tmpIndex
+		}
+	}
+	return index
+}
+
+func searchLastLessElement(nums []int64, target int64) int {
+	low, high := 0, len(nums)-1
+	for low <= high {
+		mid := low + ((high - low) >> 1)
+		if nums[mid] <= target {
+			if (mid == len(nums)-1) || (nums[mid+1] > target) {
+				// find the last less element that equal to target element
+				return mid
+			}
+			low = mid + 1
+		} else {
+			high = mid - 1
+		}
+	}
+	return -1
+}
+
+// clusterResourceNodeComparator provides a fast comparison on clusterResourceNodes
+func clusterResourceNodeComparator(a, b interface{}) int {
+	s1 := a.(ClusterResourceNode)
+	s2 := b.(ClusterResourceNode)
+	var diff int64
+	for index := 0; index < len(defaultModelSorting); index++ {
+		tmp1, tmp2 := s1.resourceList[defaultModelSorting[index]], s2.resourceList[defaultModelSorting[index]]
+		quantity1, ok := tmp1.AsInt64()
+		if !ok {
+			klog.Infof("ModelComparator: Unable to parse the values of %v's quantity1 in the cluster", defaultModelSorting[index])
+		}
+		quantity2, ok := tmp2.AsInt64()
+		if !ok {
+			klog.Infof("ModelComparator: Unable to parse the values of %v's quantity2 in the cluster", defaultModelSorting[index])
+		}
+		diff = quantity1 - quantity2
+		if diff < 0 {
```

Comment:
Now the value judgment mainly depends on the resource with the highest priority, which may lead to some misjudgments.

Reply:
You said internal sorting rule can be ignored.